File: mobile/lib/protocol/frame_builder.dart
import 'dart:typed_data';

import 'frame.dart';

/// Convenience builder for encoded frames. The session layer owns sequence
/// numbers and flags (D1/D2) and uses this instead of letting the codec decide.
/// encode still rejects reserved flags defensively.
class FrameBuilder {
  FrameBuilder._();

  static Uint8List build({
    required int messageType,
    required Uint8List payload,
    required int sequence,
    bool ackRequested = false,
    bool mustUnderstand = false,
    int versionMajor = 1,
    int versionMinor = 0,
    int timestamp = 0,
  }) {
    var flags = 0;
    if (ackRequested) {
      flags |= FrameCodec.ackRequested;
    }
    if (mustUnderstand) {
      flags |= FrameCodec.mustUnderstand;
    }
    return FrameCodec.encode(
      ProtocolFrame(
        versionMajor: versionMajor,
        versionMinor: versionMinor,
        flags: flags,
        messageType: messageType,
        sequence: sequence,
        timestamp: timestamp,
        payload: payload,
      ),
    );
  }
}

File: desktop/Protocol/FrameBuilder.cs
namespace CTRL.Desktop.Protocol;

/// <summary>
/// Convenience builder for encoded frames. The session layer uses this to keep
/// control of sequence numbers and flags (D1/D2) instead of letting the codec
/// decide. Encode still rejects reserved flags defensively.
/// </summary>
public static class FrameBuilder
{
    public static byte[] Build(
        byte messageType,
        byte[] payload,
        ushort sequence,
        bool ackRequested = false,
        bool mustUnderstand = false,
        byte versionMajor = 1,
        byte versionMinor = 0,
        ulong timestamp = 0)
    {
        var flags = (byte)0;
        if (ackRequested)
            flags |= FrameCodec.AckRequested;
        if (mustUnderstand)
            flags |= FrameCodec.MustUnderstand;

        return FrameCodec.Encode(new ProtocolFrame(
            versionMajor,
            versionMinor,
            flags,
            messageType,
            sequence,
            timestamp,
            payload));
    }
}